### ISSUE TITLE
AGS 4.0.0 Milestone proposal

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1298]
I'd like to begin setting up certain milestone for a first ags4 release, to help with prioritizing tasks. IMO there are few tasks that must be done and without which completion there's not much point in releasing ags4, at least not as a major "ags4" version. Then come things that would be very nice to have but that have less priority - I'd call these "preferable".

This list of course does not mean that other things should not be added or worked on, but whenever possible it's best to choose one of the priority tasks.
The list may be amended. Also there's now a new milestone for 4.0.0 to use for related issues.

For the reference, couple of older tickets related to ags4 planning: #450, #1121

---

### Priority tasks

 Task | Issues | Comments
 --- | --- | ---
Unicode support | #1297 | Essential to allow better multilanguage support in games
Open game project format |  #469, #1008, #1281 | For easier use with source control, teamwork, be able to manage project outside of an editor
Stand-alone build tools | [3.* tools milestone](https://github.com/adventuregamestudio/ags/milestone/10) | Got to be able to compile full game without launching an editor. NOTE: will be partially done in 3.6 branch, but some tools will be specific for ags4 after the planned project format changes.
Deploy games for Android and Mac OSX (if possible) from the Editor | TBD | There's an Android building plugin for the editor created by monkey0506 which could be integrated in the editor accompanied with proper instructions in the manual. Packing a game for OSX probably involves gathering engine binaries and game data? NOTE: also may be done in 3.6 branch, but if not for some reason then should be definitely be made in ags4

---

### Preference tasks

 Task | Issues | Comments
 --- | --- | ---
RTTI and full managed pointer support in scripts | #1259  | There's a new compiler by fernewelten with some good additions, but adding full pointer support in managed structs will resolve a multitude of scripting probems at once.
Hardware accelerated GUI controls | TBD | GUI controls are currently the only objects that are not converted to textures, but always drawn on a parent GUI surface using software method. This is inconvenient for various reasons.
Hardware accelerated DynamicSprite | TBD | Dynamic sprites are currently manipulated in software mode only, writing an implementation for 3D renderers too might enhance rendering perfomance and make further expansion of its API with new effects more interesting. But this may also require adjusting a way of how sprites are refered to and handled in the engine.

**Comment by ivan-mogilko** [comment]
As a "note to self", following tasks may be a priority for me:

1) Fix rotated GUIs. Since the gui controls are now rendered as individual textures, they have to be clipped to GUI borders using Clip/Scissor method in 3D renderers. However this does not work when GUI is rotated, as the scissor rect must be orthogonal on screen. From what I've read this may be resolved either using shaders, or stencils. Stencils, in this case, may be used to create a sort of texture mask of any shape and position. I did a small test before, and this method seem to work, but will require some enhancement to the sprite batch rendering code first (see #1708). TODO: write a ticket about this problem.
2) Script RTTI, which enables managed pointers inside managed structs (#1259). I did a [successful experiment](https://github.com/ivan-mogilko/ags-refactoring/commits/experiment--rtti) a while ago, but unfortunately got distracted by other tasks and never finished it. The implementation seem not much complicated and mostly requires cleaning up new data format and tidy the code. The experiment was done on an old compiler, but the data format should be the same with the new compiler, perhaps only the way the data is gathered during compilation may differ.

Also, if I manage to complete these, I'd look into the new sprites storage in project (#1281). I recall that @persn [started working on it](https://github.com/persn/ags/tree/feature/open-sprites), but I have not checked it yet.

**Comment by AlanDrake** [comment]
For point 2 you could render the whole GUI into a single texture, no?.
Currently, a GUI with some Transparency has its elements overlap opacities individually, instead of just the final result having that set opacity.
Maybe it could be fixed by changing alpha blend modes to replace, but still I think a GUI is best represented by a single texture.

**Comment by ivan-mogilko** [comment]
> For point 2 you could render the whole GUI into a single texture, no?.

Are you referring to making a texture a render target?

**Comment by AlanDrake** [comment]
> Are you referring to making a texture a render target

Yes, unless stencils or shaders work better. Unfortunately, I'm weak with render stuff, so I know no better than the simplicity of a texture.
I'd imagine one day we may be able to specify if certain viewports or parts can be rendered at specific resolutions, and positioned around the screen dynamically via script, if we ever get to support non-fixed game resolutions.
Maybe unrelated to this case and not important now, but should be considered for future decisions (or not).

> I recall that the idea was that the actual control's opacity = gui opacity * control opacity.

But in that case, with a transparent GUI and a control inside, you get this:
![immagine](https://user-images.githubusercontent.com/1430815/204122226-36e0b8e1-1091-48b6-a31f-46fde568463b.png)
I guess this would be solved by stencils, if I understood what I read.

**Comment by ivan-mogilko** [comment]
Opened a dedicated ticket about the GUI clipping problem: #1837

@AlanDrake
> But in that case, with a transparent GUI and a control inside, you get this:
Could you open a bug ticket, with some information about how gui is composed? I am not certain if this is a transparency bug, or a clipping bug.

**Comment by AlanDrake** [comment]
@ivan-mogilko  I opened an issue #1838 with a few screenshots

**Comment by ericoporto** [comment]
> Hardware accelerated DynamicSprite

I would postpone this (4.2), designing the API for this won't be trivial.

> Better Audio playback API

I would postpone this (4.2) too, designing the API would eat a lot of time, plus there's a suggestion of switching how sound works (for effects), let's use what we have at least in the first 4.0 release.

> Stand-alone build tools

We can advance this but switching how the editor works for this is something we should postpone for 4.2 if possible.

**Comment by ivan-mogilko** [comment]
Splitting the above list is fine. This ticket could be renamed into "priority list for 4.+", and a new one opened, with a narrow selection. It should also include tasks for testing and/or fixing existing features. One good example is custom BlendModes, that probably will have to be reimplemented as shaders to work properly (the reasons are explained in #1885).

I also may suggest that for the early development of 4.* the focus should be set on the "quality of life" improvements that affect larger scale, and larger range of users. Meanwhile it's best to not bother with minor additions, and minor issues that already have a solution or a workaround.

**Comment by ivan-mogilko** [comment]
There's no actual ticket for this yet, but might leave here for a reference: since there are few major additions planned related to the input (joystick and touch screen API), imho we should also consider reorganize existing mouse & key API and make everything similarly flexible and consistent.

For instance, instead of (or along with) `on_key_press` and `on_mouse_click` have separate `on_key_down/up` and `on_mouse_down/up`.

Also, perhaps, have an ability to detect whether a device is connected (mouse & keyboard).

In regards to a keyboard, there is a long existing problem of a non-QWERTY keyboards being more difficult to handle in script. This is briefly explained in this comment to ticket #1391:
https://github.com/adventuregamestudio/ags/issues/1391#issuecomment-1166385188
We might need to create a proper dedicated ticket for this instead.

**Comment by ivan-mogilko** [comment]
Back to this, I suppose we should beging to wrap the AGS 4 up, and decide on which *minimal* tasks out of the "big tasks" list should be a part of the 4.0 release.

I think that these two are good candidates:
* True 32-bit color values
* Update TTF renderer (use latest FreeType version)

Because both of them are improving existing and commonly used functionality, while likely break backwards compatibility.

Also @ericoporto has been working on a touch API, which will also be a nice addition, would it be done in time.

**Comment by ericoporto** [comment]
The touch "API" is working, the issue is not the API itself, it's multi-touch support on GUIs. :/

**Comment by ivan-mogilko** [comment]
> The touch "API" is working, the issue is not the API itself, it's multi-touch support on GUIs. :/

Is the GUI blocking touch API, or is it a side problem? Perhaps you could add touch API first to let users use that in script, and leave GUI problem for the later?

**Comment by ivan-mogilko** [comment]
Another checkpoint... 32-bit drawing colors are now supported. So my updated proposal on a minimal list to get to this version is:

* Update TTF renderer (use latest FreeType version)
* Touch API; there's a Draft #2440, but I remember that I had doubts about the API. I'd like to check it out again, hopefully will be able to do so on the upcoming week.

Besides these, I think that following will be good to do too, if possible:
* Parent hierarchy for the script classes (see my dirty technical experiment here: https://github.com/adventuregamestudio/ags/issues/1228#issuecomment-2657947046).
* Dynamic cast of pointers in script (see #2665).

**Comment by ivan-mogilko** [comment]
Update:
* Touch API implemented in #2692
* Dynamic pointer cast in script implemented in #2665 and #2693

**Comment by ericoporto** [comment]
Hey, I would like to make a new issue with a todo list of what we need to do to move from alpha to beta in ags4, so that we can later think about moving onto RC and all… I feel most of the major pain points are solved but the lingering stuff is not well defined - I remember I couldn’t properly finish the joystick controls, like, I was out of ideas for the things that we don’t script and how to configure the default button of the joystick for things like skipping dialog, video, …

Probably also doing a recheck of the issues that are marked as ags4 to see what is really necessary for this release - I guess a sibling issue should be in the manual repository and other in the templates?

**Comment by ivan-mogilko** [comment]
> Hey, I would like to make a new issue with a todo list of what we need to do to move from alpha to beta in ags4

Right, there've been a lot of ideas about 4.0 that unfortunately never got addressed, but it's taking way too long already, so better to wrap it up into the beta and then final release.

We might open the AGS 4 Beta ticket with the list.

I think, at the minimal, these tasks should be done for the Beta:
1. SDL_ttf library for the fonts rendering, with potentially adding BiDi and Harfbuzz (latter may be postponed). I remember there were concerns about compatibility with some classic AGS fonts, and performance optimization, but these may be solved during Beta stage, or even further in patches.
2. AudioPlayer class, as a one-time playback instance, to be returned when starting a clip playback. I spoke about this on forums, and I believe this may be achieved with relative ease, transferring parts of AudioChannel API.
3. Dialog scripts as separate files (you have it in works now, right #2978), and dialogs themselves if you feel like achieving this along (but dialog scripts is a minimal requirement here).
4. Review the new script API and make adjustments if seems necessary.
5. Consider parent class(es) in script. I had lots of doubts about this one, one time thinking that i might do something, and then being in doubts again. Recently, I've been thinking about adding at least 1 ultimate managed class called `Anything` (yes) with only 1 function `GetTypeName()`, that could be a base class for every non-static managed class in script, and even supporting being used as user class parent. Because we now support dynamic pointer downcast in scripts, such base class could be used to at least store multiple types in one array.

**Comment by ericoporto** [comment]
> SDL_ttf library for the fonts rendering … I remember there were concerns about compatibility

I think the issue was in the font metrics, it wasn’t trivial to retrieve the same information we have, but I think if we have our own fork of it like you did in your PR this is trivial to fix.

> Review the new script API and make adjustments if seems necessary.

I don’t know if it’s this, but there are a few functions that don’t “belong” to anything (global?), that could perhaps be moved. SetAmbientLightLevel and SerAmbientTint could be moved to the Room object, SetMultitaskingMode could be moved to System, and perhaps made into a property. I can create an issue with a few suggestions. This is mostly “sugar”, with nothing actually changing, just trying to reduce the global functions.

> Anything

Is this for all objects or just the ones that have a visual representation? (Like AudioClip doesn’t have a X and Y position, …)

**Comment by ivan-mogilko** [comment]
> > SDL_ttf library for the fonts rendering … I remember there were concerns about compatibility

> I think the issue was in the font metrics, it wasn’t trivial to retrieve the same information we have, but I think if we have our own fork of it like you did in your PR this is trivial to fix.

No, the compatibility issue that I refer to is that some fonts appear garbled. I had a guess that this is related to hinting, but never had a time to experiment with this.

> > Review the new script API and make adjustments if seems necessary.

> I don’t know if it’s this, but there are a few functions that don’t “belong” to anything (global?), that could perhaps be moved. 

I was referring to the new script API, that has to be reviewed once more before it's "solidified". Replacing old api with a new one can be done later, even during Beta stage, since this goes through a common api deprecation process.

>> Anything

> Is this for all objects or just the ones that have a visual representation? (Like AudioClip doesn’t have a X and Y position, …)

`Anything` is for all objects. The idea is to have a pointer type that could point to every managed class. It only has 1 member that returns its type name, and nothing else, therefore suits as a parent to any possible class.

**Comment by ericoporto** [comment]
> Anything is for all objects. The idea is to have a pointer type that could point to every managed class. It only has 1 member that returns its type name, and nothing else, therefore suits as a parent to any possible class.

This idea is pretty magical in the possibilities it opens. I think everything managed has a type name - like, even stuff in plugins! Is the idea here to have everything implicitly inherit it (or have the Type property like the Length one was made), or the idea would still use the `extends` keyword?

It sounds to me like the object type in C# except Anything is a better AGS name - it is more obvious and the word object already has a different meaning in AGS world.

**Comment by ivan-mogilko** [comment]
> This idea is pretty magical in the possibilities it opens. I think everything managed has a type name - like, even stuff in plugins! Is the idea here to have everything implicitly inherit it (or have the Type property like the Length one was made), or the idea would still use the extends keyword?

I would love to see it be inherited automatically, but that would require a new compiler feature. For starters, I would just do this manually for engine API, and leave users to extending their classes if they wish to.

Internally, TypeName can be implemented by calling existing GetType() for engine's and plugins's dynamic objects, and retrieving the object's typeinfo from RTTI for user script classes.

**Comment by ericoporto** [comment]
Just thinking about this, supposedly somewhere in the compiler it parses the extends keyword and handles inheritance from whatever type it is being inherited.

Now somewhere else in the compiler it reads a managed type declaration and does… Something. So if it was a nice function (doubt, but let’s pretend the code is organized), we could then always also run the function that handles the extends keyword with the type “Anything” hardcoded there. Edit: I think the place is in the if clause that handles the SYM_EXTENDS, we would add an else clause to it and then it could be worked out there, if it doesn't extend anything, it would at least extend `Anything` if this is a managed struct.

I guess in C# [object is a language feature](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/types#823-the-object-type) so it is probably a keyword there. This makes me feel like it is alright to have it hardcoded in the compiler.

If we still need Anything struct declared in agsdefn.sh then we would also need to handle to not let it inherit itself.

Edit: actually, I think we can just throw it in the symbol table when the symbol table is initialized somehow.

Edit2: cc @fernewelten

**Comment by ivan-mogilko** [comment]
I'd rather not discuss this here, but open a different ticket for this.

But I suppose that we could use a keyword for declaring a "automatic managed parent" and use such keyword in `Anything` struct's declaration.

There's a similar keyword called "internalstring", which is used when declaring a String, although i forgot what it does in practice.

It's best to declare the struct itself in the script IMO, because that makes it easier to edit one if necessary, otherwise we'd have to hack the compiler every time it needs a change, and that's quite bad.

**Comment by ivan-mogilko** [comment]
Following issues are better done prior to Beta too.

General deprecation of things:
#2553

Related to new room format in templates:
#2561
#2562

Changing parameter meaning, may break existing scripts:
#2502 (either apply or cancel)
#2549
#2856

**Comment by ivan-mogilko** [comment]
Opened a Alpha->Beta transition checklist here: #3019.

Will close this, as sort-of accomplished.
